### PR TITLE
chore: remove support for python 3.9

### LINKED ISSUE #818: Drop Python 3.9 on `kedro-datasets`
## Description
Following NEP 29, let's drop Python 3.9 on `kedro-datasets`. NumPy has already done so. https://numpy.org/doc/stable/release/2.1.0-notes.html

In #695 there was a proposal to review the current policy, but nothing happened. It comes from https://github.com/kedro-org/kedro-plugins/pull/689, where we expressed some reservations about the potential impact on Viz. Some months have passed though, so hopefully it's safe to do now.

### PR DESCRIPTION
**Pull request by felixscherz**
## Description
Hi, this is removing support for python 3.9 and updates the python syntax accordingly. Intended to resolve #818. 

## Development notes
- removed `typing.Union` in favor of `|` operator
- removed `typing.Callable` in favor of `collections.abc.Callable`
- removed python 3.9 from unit test matrix in CI

## Checklist

- [ ] Opened this PR as a 'Draft Pull Request' if it is work-in-progress
- [x] Updated the documentation to reflect the code changes
- [x] Added a description of this change in the relevant `RELEASE.md` file
- [ ] Added tests to cover my changes

### CODE REVIEW

**Comment by deepyaman** [comment]
I think can also simplify https://github.com/felixscherz/kedro-plugins/blob/chore/remove-python-3.9/kedro-datasets/pyproject.toml#L230-L231; not sure if there are other such places.

On a separate note (and this is in no way a reflection on this PR, which resolves the issue very well), I've started thinking more about https://github.com/kedro-org/kedro-plugins/issues/695, especially

> Is it necessary to drop 3.9 support when it's compatible and not causing any issue?

While I was a big proponent of NEP-29, [some experience on another project](https://github.com/ibis-project/ibis/issues/10064) makes me wonder if we should hold off a bit, especially given almost nothing of value is gained through these changes (the typing changes are nice, but don't have any real impact...). I wouldn't be opposed to holding off a major version, unless somebody can point to more bad things we're doing right now to support 3.9.

**Comment by felixscherz** [comment]
Thanks for having a look, good catch with the environment markers! I'll remove the one you referenced which seems to be the only one that deals with python 3.9.
I would agree with you that removing support is not actually necessary as long as there are no major workarounds required and python 3.9 is still officially supported.

File: kedro-datasets/RELEASE.md [hunk 1]
```diff
@@ -1,5 +1,6 @@
 # Upcoming Release
 ## Major features and improvements
+* Removed support for Python 3.9
```
Comment (merelcht) [inline kedro-datasets/RELEASE.md:3]: Don't forget to add your name to the contributor list 🙂

**Review by merelcht — APPROVED** [review]
Thanks for the contribution @felixscherz ! I'm still in favour of adhering to NEP-29, which is what we agreed on recently in backlog grooming as well. 

@astrojuanlu I'm guessing you also still want to go ahead with this?

File: kedro-datasets/RELEASE.md (hunk 1 above)
Comment (felixscherz) [reply kedro-datasets/RELEASE.md:3]: Will do:)

**Review by astrojuanlu — APPROVED** [review]
Agreed to continue adhering to NEP 29 indeed! Thanks @felixscherz for this PR 🙌🏼 

Approving to signal my acceptance although naturally the CI will need to be green

**Comment by felixscherz** [comment]
Thanks for the review! I'm having a look at the CI failures, it seems like they are failing on the `main` branch as well.

**Comment by kevin1kevin1k** [comment]
@felixscherz Hey I am the author of https://github.com/kedro-org/kedro/pull/4212.
I noticed that under `.github/workflows` you modified `kedro-datasets.yml`.
According to the error patterns I guess the CI errors will be gone after modifying `kedro-{airflow, docker, telemetry}.yml` as well.

**Comment by felixscherz** [comment]
@kevin1kevin1k thank you! I opened an issue to deal with the removal of python 3.8: #872, we need to remove it from github actions but also make updates to the python syntax and update the `pyrpoject.toml` files, I'm working on those PRs at the moment:)

**Pull request merged**